File: src/com/javascript-practice/linkList/removeNode.js

const removeNodeRecursive = (node, target) => {

    if(node === null) return null;
    if(node.val === target) 
    return node.next;
    
    node.next = removeNodeRecursive(node.next, target);
    return node;

}



const removeNode = (node, target) => {

    if(node.val === target) return node.next;

    let current = node ;
    let prev = null;
    while(current!=null) {

        if(current.val === target) {
            prev.next = current.next;
            break;
        }
         prev = current;
         current = current.next;

    }
    return node;
}

class Node {
    constructor(val) {
        this.val  = val;
        this.next = null;
    }
}
const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");

a.next = b;
b.next = c;
c.next = d;
d.next = e;
e.next = f;

// a -> b -> c -> d -> e -> f

// let result  = removeNode(a, "c");
let result  = removeNodeRecursive(a, "b");
// a -> b -> d -> e -> f
while(result!==null) {
    console.log(result.val);
    result = result.next;
}

File: src/com/structy/LinkedList/RemoveNode.java
package com.structy.LinkedList;

public class RemoveNode {

    public Node removeNode(Node head, String target){
        if(target.equals(head.strVal)){
            return head.next;
        }
        Node current = head;
        Node prev = null;
        while (current!=null){
            if(target.equals(current.strVal)){
               prev.next = current.next;
               //Need to break for another removal
                break;
            }
            prev = current;
            current = current.next;
        }
        return head;
    }

    public Node removeNodeRec(Node head, String target){
        if(head == null)
            return null;

        if(target.equals(head.strVal)){
            return head.next;
        }
        head.next = removeNodeRec(head.next, target);
        return head;
    }

    public static void main(String[] args) {
        Node a = new Node("a");
        Node b = new Node("b");
        Node c = new Node("c");
        Node d = new Node("d");
        Node e = new Node("e");
        Node f = new Node("f");

        a.next = b;
        b.next = c;
        c.next = d;
        d.next = e;
        e.next = f;

        RemoveNode removeNode = new RemoveNode();
        Node result  = removeNode.removeNodeRec(a, "f");
        Node tmp = result;
        while (tmp!=null){
            System.out.print(tmp.strVal);
            tmp = tmp.next;
        }
    }
}
